lib/datadog/grok/build.rs

```diff
@@ -0,0 +1,41 @@
+extern crate lalrpop;
+
+use std::fmt::Write as fmt_write;
+use std::fs::File;
+use std::io::{BufRead, BufReader};
+use std::path::Path;
+use std::{env, fs};
+
+fn main() {
+    lalrpop::Configuration::new()
+        .always_use_colors()
+        .process_current_dir()
+        .unwrap();
+
+    println!("cargo:rerun-if-changed=src/parser.lalrpop");
+
+    read_grok_patterns();
+}
+
+/// Reads grok patterns defined in the `patterns` folder into the static `PATTERNS` variable
+fn read_grok_patterns() {
+    let mut output = "static PATTERNS: &[(&str, &str)] = &[\n".to_string();
+
+    fs::read_dir(Path::new("patterns"))
+        .expect("can read 'patterns' dir")
+        .filter_map(|path| File::open(path.expect("can read 'patterns' dir").path()).ok())
+        .flat_map(|f| BufReader::new(f).lines().filter_map(|l| l.ok()))
+        .filter(|line| !line.starts_with('#') && !line.is_empty())
+        .for_each(|line| {
+            let (key, value) =
+                line.split_at(line.find(' ').expect("pattern is 'ruleName definition'"));
+            write!(output, "\t(\"{}\", r#\"{}\"#),", key, &value[1..])
+                .expect("can append patterns");
```

Comment:
Nit: I've noticed that your `.expect` messages tend to include what _should_ be expected (which makes sense, given the name of the function!).

On the whole, we've tended to favour the approach of stating _what went wrong_ so these make for saner panic messages (IMO, it looks a little odd to see "thread 'main' panicked at: can append patterns" if the reason it panicked is because patterns _couldn't_ be appended.)

This models the [Rust book format](https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html) a little closer, I think.

Doing a cursory search, I noticed this isn't always the case -- e.g. we do have a few examples such as:

```
.expect("validated enum")
```

But I think it makes sense to dump these out in the context of 'what went wrong', since the fact that this was an expect and not a panic isn't obvious to the user.

Reply:
(I don't think this has to hold up merging the PR, btw -- just something I noticed!)

Reply:
Yeah, totally makes sense. The function name is a bit confusing indeed.